dumpling/export/conn.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package export
+
+import (
+	"database/sql"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tidb/br/pkg/utils"
+	tcontext "github.com/pingcap/tidb/dumpling/context"
+)
+
+// BaseConn wraps connection instance.
+type BaseConn struct {
+	DBConn *sql.Conn
+
+	backOffer     backOfferResettable
+	rebuildConnFn func(*sql.Conn, bool) (*sql.Conn, error)
+}
+
+func newBaseConn(conn *sql.Conn, shouldRetry bool, rebuildConnFn func(*sql.Conn, bool) (*sql.Conn, error)) *BaseConn {
+	baseConn := &BaseConn{DBConn: conn}
+	baseConn.backOffer = newRebuildConnBackOffer(shouldRetry)
+	if shouldRetry {
+		baseConn.rebuildConnFn = rebuildConnFn
+	}
+	return baseConn
+}
+
+// QuerySQL defines query statement, and connect to real DB.
+func (conn *BaseConn) QuerySQL(tctx *tcontext.Context, handleOneRow func(*sql.Rows) error, reset func(), query string, args ...interface{}) error {
+	retryTime := 0
+	err := utils.WithRetry(tctx, func() (err error) {
+		retryTime++
+		if retryTime > 1 && conn.rebuildConnFn != nil {
+			conn.DBConn, err = conn.rebuildConnFn(conn.DBConn, false)
+			if err != nil {
+				return
+			}
+		}
+		err = simpleQueryWithArgs(conn.DBConn, handleOneRow, query, args...)
```

Comment:
because we pass tctx now, so `simpleQueryWithArgs` better not use background ctx